lib/SILOptimizer/FunctionSignatureTransforms/ExistentialTransform.cpp

```diff
@@ -0,0 +1,595 @@
+                : IsTake_t::IsTake,
+            IsInitialization_t::IsInitialization);
+        if (ExistentialArgDescriptor[ArgDesc.Index].DestroyAddrUse) {
+          NewFBuilder.createDestroyAddr(InsertLoc, NewArg);
+        } else {
+          MissingDestroyUse = true;
+        }
+        ValueMap.insert(std::make_pair(OrigArg, SILValue(ASI)));
+        break;
+      }
+      case ExistentialRepresentation::Class: {
+        ///  Simple case: Create an init_existential.
+        /// %5 = init_existential_ref %0 : $T : $T, $P
+        auto *InitRef = NewFBuilder.createInitExistentialRef(
+            InsertLoc, ArgDesc.Arg->getType(),
+            NewArg->getType().getASTType()->getCanonicalType(), NewArg,
+            Conformances);
+        ValueMap.insert(std::make_pair(OrigArg, SILValue(InitRef)));
+        break;
+      }
+      default: {
+        llvm_unreachable("Unhandled existential type in ExistentialTransform!");
+        break;
+      }
+      };
+    } else {
+      /// Arguments that are not rewritten.
+      auto Ty = params[ArgDesc.Index].getType();
+      if (!(Ty->hasTypeParameter() || Ty->hasArchetype())) {
+        NewArg = ClonedEntryBB->createFunctionArgument(ArgDesc.Arg->getType(),
+                                                       ArgDesc.Decl);
+        NewArg->setOwnershipKind(ValueOwnershipKind(
+            M, ArgDesc.Arg->getType(), ArgDesc.Arg->getArgumentConvention()));
+      } else {
+        SILType GenericSILType =
+            M.Types.getLoweredType(NewF.mapTypeIntoContext(Ty));
+        NewArg = ClonedEntryBB->createFunctionArgument(GenericSILType);
+        NewArg->setOwnershipKind(ValueOwnershipKind(
+            M, GenericSILType, ArgDesc.Arg->getArgumentConvention()));
+      }
+      ValueMap.insert(std::make_pair(OrigArg, SILValue(NewArg)));
+    }
+  }
+
+  // Recursively visit original BBs in depth-first preorder, starting with the
+  // entry block, cloning all instructions other than terminators.
+  visitSILBasicBlock(OrigEntryBB);
+
+  // Now iterate over the BBs and fix up the terminators.
+  for (auto BI = BBMap.begin(), BE = BBMap.end(); BI != BE; ++BI) {
+    NewFBuilder.setInsertionPoint(BI->second);
+    visit(BI->first->getTerminator());
+  }
+
+  /// If there is an argument with no DestroyUse, insert DeallocStack
+  /// before return Instruction.
+  llvm::SmallPtrSet<ReturnInst *, 4> ReturnInsts;
+  if (MissingDestroyUse) {
+    /// Find the set of return instructions in a function.
+    for (auto &BB : NewF) {
+      TermInst *TI = BB.getTerminator();
+      if (auto *RI = dyn_cast<ReturnInst>(TI)) {
+        ReturnInsts.insert(RI);
+      }
+    }
+  }
+
+  /// Walk over destroy_addr instructions and perform fixups.
+  for (auto &ArgDesc : reversed(ArgumentDescList)) {
+    int ArgIndex = ArgDesc.Index;
+    auto iter = Arg2AllocStackMap.find(ArgIndex);
+    if (iter != Arg2AllocStackMap.end()) {
+      auto *ASI = dyn_cast<AllocStackInst>(iter->second);
```

Comment:
It looks like this cast should never fail, because you dereference ASI below. Can you change the type of the DenseMap so that the value is already an AllocStackInst?